### PR TITLE
Fix nth-last-child styles on virtualizing layouts

### LINKED ISSUE #9997: "nth-last-child" is getting broken (not applying on the elements) after you scroll with scrollbar thumb.
**Describe the bug**

"nth-last-child" is getting broken (not applying on the elements) after you scroll with scrollbar thumb.

**To Reproduce**

1. ListBox page, control catalog
2. Scroll with scrollbar thumb.
3. See error.
4. Scroll with a wheel.
5. Error is gone.

**Desktop (please complete the following information):**

- OS: Windows
- Version: master, right after https://github.com/AvaloniaUI/Avalonia/pull/9677 merged

https://user-images.githubusercontent.com/3163374/213033809-c635b6f6-0ae5-4f6a-a5b6-eb031774dab1.mp4

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

After #9677 was merged, `nth-last-child` styles stopped working in `ListBox` due [to the problem described here](https://github.com/AvaloniaUI/Avalonia/issues/9997#issuecomment-1398266638).

This PR modifies the `IChildIndexProvider` API to add the following:

- The new index of the child is now passed in `ChildIndexChangedEventArgs`
- A `TotalCountChanged` event was added

It also removes the hack in `ItemsControl` which was used to work around this problem.

## Questions

I've marked this PR as draft for the moment because I'm not sure about one thing:

We now have two events on `IChildIndexProvider` but they will always be subscribed by the same object. Maybe we need to merge the two events into a single event with an `Action` property which would be an enum of:

- `ChildIndexChanged`
- `TotalCountChanged`

## Breaking changes

The `IChildIndexProvider` interface has changed. Shouldn't be much of a problem as it's unlikely to be client-implemented.

## Fixed issues

Fixes #9997

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0028939-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

File: src/Avalonia.Base/Styling/Activators/NthChildActivator.cs [hunk 1]
```diff
@@ -28,24 +30,50 @@ internal sealed class NthChildActivator : StyleActivatorBase
...
-        protected override void Deinitialize() => _provider.ChildIndexChanged -= ChildIndexChanged;
+        protected override void Initialize()
+        {
+            _provider.ChildIndexChanged += ChildIndexChanged;
+            _provider.TotalCountChanged += TotalCountChanged;
+        }
+
+        protected override void Deinitialize()
+        {
+            _provider.ChildIndexChanged -= ChildIndexChanged;
```
Comment (maxkatz6) [inline src/Avalonia.Base/Styling/Activators/NthChildActivator.cs:45]: 1. Do we even need to subscribe on TotalCountChanged when it's not reversed?
2. TotalCountChanged  is not unsubscribed on Deinitialize

Comment (grokys) [reply src/Avalonia.Base/Styling/Activators/NthChildActivator.cs:45]: 1. I think this depends on the answer to the question in the PR description
2. This too, really..,

**Comment by grokys** [comment]
@maxkatz6 ping - still waiting for your thoughts on the question in the PR description.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0030360-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

File: src/Avalonia.Controls/Panel.cs [hunk 2]
```diff
@@ -161,6 +170,12 @@ private protected virtual void InvalidateMeasureOnChildrenChanged()
             InvalidateMeasure();
         }
 
+        private void ChildrenPropertyChanged(object? sender, PropertyChangedEventArgs e)
+        {
+            if (e.PropertyName == nameof(Children.Count) || e.PropertyName is null)
+                _totalCountChanged?.Invoke(this, EventArgs.Empty);
+        }
```
Comment (maxkatz6) [inline src/Avalonia.Controls/Panel.cs:177]: There is a lot of panels used in the application. Would it make sense to subscribe on Children.PropertyChanged only when TotalCountChanged has any subscription? Could be a minor one though.

**Comment by maxkatz6** [comment]
> We now have two events on IChildIndexProvider but they will always be subscribed by the same object. Maybe we need to merge the two events into a single event with an Action property which would be an enum.

I don't have strong preference here. Having a single event makes sense in general.

If keep two different events, then total count can be added to the TotalCountChanged event args. So TryGetTotalCount/GetChildIndex calls can be basically eliminated except of initial read.
But I think it's a hardly beneficial, as it will add allocations for an additional argument for each item added.
Without that, there is not much usage of independent TotalCountChanged.

File: src/Avalonia.Controls/Panel.cs (hunk 2 above)
Comment (grokys) [reply src/Avalonia.Controls/Panel.cs]: Fixed.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0030423-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Review by maxkatz6 — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0030445-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Pull request merged**